---
title: "Examine_SV_20220105"
author: "Carissa Sherman"
date: "2023-01-06"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
##Background##
In this code, I'm comparing our phenotype frequencies to tables from PharmGKB: https://www.pharmgkb.org/page/pgxGeneRef. Our phenotypes are generated from genotype-phenotype translation tables from CPIC. There may not be comparison for all of the genes available on PharmGKB. 

##Libraries##
```{r}
library(xlsx)
library(dplyr)
library(tidyr)
library(writexl)
library(stringr)
library(readxl)
library(readr)
library(openxlsx)
```

##Phenotypes##
```{r}
#Phenotypes from the 1KGP
phenotypes <- read.csv("Phenotype Frequencies/SupplementaryTable4_PhenotypeFrequencies.csv")

#Genes available from PharmGKB with allele and phenotype frequencies: https://www.pharmgkb.org/page/pgxGeneRef
genes = c("CYP2B6","CYP2C9","CYP2C19","CYP2D6","CYP3A5","DPYD","NUDT15","TPMT","UGT1A1","ABCG2","SLCO1B1")

#Subsetting 1KGP data with genes availabile from PharmGKB
phenotypes = phenotypes[phenotypes$Gene %in% genes,]

pheno_gene = phenotypes %>% 
  group_split(Gene)

names(pheno_gene) = unique(phenotypes$Gene)
```

#ABCG2
```{r}
#Reformat the dataset
ABCG2 = pivot_wider(pheno_gene[["ABCG2"]], names_from = "Superpopulation.code", values_from = Frequency)

ABCG2_lit = read_excel("Phenotype Frequencies/ABCG2_frequency_table.xlsx", sheet = 4) 

ABCG2_and_lit = merge(ABCG2, ABCG2_lit, by = "Phenotype")

#amr_lit, sas_lit, eas_lit, eur_lit, AFR_lit
col_order = c("Phenotype", "Gene","AFR","AFR_lit","EAS","EAS_lit","EUR","EUR_lit","SAS","SAS_lit","AMR","AMR_lit")

ABCG2_and_lit = ABCG2_and_lit[, col_order]

ABCG2_and_lit_na_rm = na.omit(ABCG2_and_lit)
```

#CYP2B6
```{r}
#reformat the data set
CYP2B6 = pivot_wider(pheno_gene[["CYP2B6"]], names_from = "Superpopulation.code", values_from = Frequency)

CYP2B6_lit = read_excel("Phenotype Frequencies/CYP2B6_frequency_table.xlsx", sheet = 4) 

CYP2B6_and_lit = merge(CYP2B6, CYP2B6_lit, by = "Phenotype")
col_order = c("Phenotype", "Gene","AFR","AFR_lit","EAS","EAS_lit","EUR","EUR_lit","SAS","SAS_lit","AMR")

CYP2B6_and_lit = CYP2B6_and_lit[, col_order]

CYP2B6_and_lit_na_rm = na.omit(CYP2B6_and_lit)
```

#CYP2C19
```{r}
#Reformat the dataset
CYP2C19 = pivot_wider(pheno_gene[["CYP2C19"]], names_from = "Superpopulation.code", values_from = Frequency)

CYP2C19_lit = read_excel("Phenotype Frequencies/CYP2C19_frequency_table.xlsx", sheet = 4) 

CYP2C19_and_lit = merge(CYP2C19, CYP2C19_lit, by = "Phenotype")
col_order = c("Phenotype", "Gene","AFR","AFR_lit","EAS","EAS_lit","EUR","EUR_lit","SAS","SAS_lit","AMR","AMR_lit")

CYP2C19_and_lit = CYP2C19_and_lit[, col_order]

CYP2C19_and_lit_na_rm = na.omit(CYP2C19_and_lit)
```

#CYP2C9
```{r}
#Reformat the dataset
CYP2C9 = pivot_wider(pheno_gene[["CYP2C9"]], names_from = "Superpopulation.code", values_from = Frequency)

CYP2C9_lit = read_excel("Phenotype Frequencies/CYP2C9_frequency_table.xlsx", sheet = 4) 

CYP2C9_and_lit = merge(CYP2C9, CYP2C9_lit, by = "Phenotype")
col_order = c("Phenotype", "Gene","AFR","AFR_lit","EAS","EAS_lit","EUR","EUR_lit","SAS","SAS_lit","AMR","AMR_lit")

CYP2C9_and_lit = CYP2C9_and_lit[, col_order]

CYP2C9_and_lit_na_rm = na.omit(CYP2C9_and_lit)
```

#CYP2D6 
```{r}
#Reformat the dataset
CYP2D6 = pivot_wider(pheno_gene[["CYP2D6"]], names_from = "Superpopulation.code", values_from = Frequency)

CYP2D6_lit = read_excel("Phenotype Frequencies/CYP2D6_frequency_table.xlsx", sheet = 5)

CYP2D6_and_lit = merge(CYP2D6, CYP2D6_lit, by = "Phenotype")
col_order = c("Phenotype", "Gene","AFR","AFR_lit","EAS","EAS_lit","EUR","EUR_lit","SAS","SAS_lit","AMR","AMR_lit")

CYP2D6_and_lit = CYP2D6_and_lit[, col_order]

CYP2D6_and_lit_na_rm = na.omit(CYP2D6_and_lit)

#write.xlsx(CYP2D6_and_lit_na_rm, "CYP2D6_20230214.xlsx")
```

#CYP3A5
```{r}
#Reformat the dataset
CYP3A5 = pivot_wider(pheno_gene[["CYP3A5"]], names_from = "Superpopulation.code", values_from = Frequency)

CYP3A5_lit = read_excel("Phenotype Frequencies/CYP3A5_frequency_table.xlsx", sheet = 4) 

CYP3A5_and_lit = merge(CYP3A5, CYP3A5_lit, by = "Phenotype")

#sas_lit, eas_lit, eur_lit, afr_lit
col_order = c("Phenotype", "Gene","AFR","AFR_lit", "EAS","EAS_lit","EUR","EUR_lit","SAS","SAS_lit","AMR")

CYP3A5_and_lit = CYP3A5_and_lit[, col_order]

CYP3A5_and_lit_na_rm = na.omit(CYP3A5_and_lit)
```

#DPYD
```{r}
#Reformat the dataset
DPYD = pivot_wider(pheno_gene[["DPYD"]], names_from = "Superpopulation.code", values_from = Frequency)

DPYD_lit = read_excel("Phenotype Frequencies/DPYD_frequency_table.xlsx", sheet = 4) 

DPYD_and_lit = merge(DPYD, DPYD_lit, by = "Phenotype")
col_order = c("Phenotype", "Gene","AFR","AFR_lit","EAS","EAS_lit","EUR","EUR_lit","SAS","SAS_lit","AMR")

DPYD_and_lit = DPYD_and_lit[, col_order]

DPYD_and_lit_na_rm = na.omit(DPYD_and_lit)
```

#NUDT15
```{r}
#Reformat the dataset
NUDT15 = pivot_wider(pheno_gene[["NUDT15"]], names_from = "Superpopulation.code", values_from = Frequency)

NUDT15_lit = read_excel("Phenotype Frequencies/NUDT15_frequency_table.xlsx", sheet = 4) 

NUDT15_and_lit = merge(NUDT15, NUDT15_lit, by = "Phenotype")

#sas_lit, eas_lit, eur_lit
col_order = c("Phenotype", "Gene","AFR","EAS","EAS_lit","EUR","EUR_lit","SAS","SAS_lit","AMR")

NUDT15_and_lit = NUDT15_and_lit[, col_order]

NUDT15_and_lit_na_rm = na.omit(NUDT15_and_lit)
```

#SLCO1B1 
```{r}
#Reformat the dataset
SLCO1B1 = pivot_wider(pheno_gene[["SLCO1B1"]], names_from = "Superpopulation.code", values_from = Frequency)

SLCO1B1_lit = read_excel("Phenotype Frequencies/SLCO1B1_frequency_table.xlsx", sheet = 4) 

SLCO1B1_and_lit = merge(SLCO1B1, SLCO1B1_lit, by = "Phenotype")

col_order = c("Phenotype", "Gene","AFR","AFR_lit","EAS","EAS_lit","EUR","EUR_lit","SAS","SAS_lit","AMR","AMR_lit")

SLCO1B1_and_lit = SLCO1B1_and_lit[, col_order]

SLCO1B1_and_lit_na_rm = na.omit(SLCO1B1_and_lit)
```

#TPMT
```{r}
#Reformat the dataset
TPMT = pivot_wider(pheno_gene[["TPMT"]], names_from = "Superpopulation.code", values_from = Frequency)

TPMT_lit = read_excel("Phenotype Frequencies/TPMT_frequency_table.xlsx", sheet = 4) 

TPMT_and_lit = merge(TPMT, TPMT_lit, by = "Phenotype")

col_order = c("Phenotype", "Gene","AFR","AFR_lit","EAS","EAS_lit","EUR","EUR_lit","SAS","SAS_lit","AMR")

TPMT_and_lit = TPMT_and_lit[, col_order]

TPMT_and_lit_na_rm = na.omit(TPMT_and_lit)
```

#UGT1A1
```{r}
#Reformat the dataset
UGT1A1 = pivot_wider(pheno_gene[["UGT1A1"]], names_from = "Superpopulation.code", values_from = Frequency)

UGT1A1_lit = read_excel("Phenotype Frequencies/UGT1A1_frequency_table.xlsx", sheet = 4) 

UGT1A1_and_lit = merge(UGT1A1, UGT1A1_lit, by = "Phenotype")

#sas_lit, eas_lit, eur_lit, afr_lit
col_order = c("Phenotype", "Gene","AFR","AFR_lit", "EAS","EAS_lit","EUR","EUR_lit","SAS","SAS_lit","AMR")

UGT1A1_and_lit = UGT1A1_and_lit[, col_order]

UGT1A1_and_lit_na_rm = na.omit(UGT1A1_and_lit)
```


#Export Datasets 
```{r}
dataset_names <- list('ABCG2' = ABCG2_and_lit_na_rm, 'CYP2B6' = CYP2B6_and_lit, 'CYP2C19' = CYP2C19_and_lit, 'CYP2C9' = CYP2C9_and_lit, 'CYP2D6' = CYP2D6_and_lit,'CYP3A5' = CYP3A5_and_lit_na_rm, 'DPYD' = DPYD_and_lit, 'NUDT15' = NUDT15_and_lit_na_rm, 'SLCO1B1' = SLCO1B1_and_lit, 'TPMT' = TPMT_and_lit, 'UGT1A1' = UGT1A1_and_lit_na_rm)

#write.xlsx(dataset_names, file = 'phenotype_frequency_comparison_2023XXXX.xlsx')
```

```{r}
sessionInfo()
```

